### PR TITLE
Capitalization Collation Tests

### PR DESCRIPTION
**Pull request by RubenCerna2079**
## Why make this change?

Solves issue #2350, to add collation tests related to capitalization.

## What is this change?

Added collation tests related to case sensitivity for `MSSQL`, `MYSQL`, & `POSTGRESQL` by creating a new `GraphQLCollationTests` file under `SQLTests` that will have all the tests related to changes in collations for SQL databases.

## How was this tested?

- [X] Integration Tests
- [ ] Unit Tests

### CODE REVIEW

File: src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+        #region Tests
+        /// <summary>
+        /// MsSql Capitalization Collation Tests
+        /// </summary>
+        [DataTestMethod]
+        [DataRow("books", "title", @"SELECT title FROM books ORDER BY title ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        [DataRow("books", "title", @"SELECT title FROM BOOKS ORDER BY TITLE ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        [DataRow("books", "title", @"SELECT title FROM Books ORDER BY Title ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        public async Task MsSqlCapitalizationResultQuery(string type, string item, string dbQuery)
```
Comment (abhishekkumams) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs:33]: Does it give expected results when we use title/TITLE/Title in graphql query as well?

Comment (abhishekkumams) [reply src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs:33]: would be good to add those as well.

**Review by abhishekkumams — APPROVED** [review]
looks good to me, Just one suggestions for adding few more data rows.

File: src/Service.Tests/SqlTests/GraphQLCollationTests/GraphQLCollationTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Azure.DataApiBuilder.Service.Tests.SqlTests.GraphQLCollationTests
+{
+    [TestClass]
+    public class GraphQLCollationTestBase : SqlTestBase
+    {
+        #region Tests
+
+        /// <summary>
+        /// Compares SQL Database Query with GraphQL Query
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/GraphQLCollationTestBase.cs:16]: Compares the queries to detect what? 
The description should try to answer "What is the goal of the test and how/why does this test achieve that goal?"

File: src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        [ClassInitialize]
+        public static async Task SetupAsync(TestContext context)
+        {
+            DatabaseEngine = TestCategory.MSSQL;
+            await InitializeTestFixture();
+        }
+
+        #region Tests
+        /// <summary>
+        /// MsSql Capitalization Collation Tests
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs:27]: what do you mean by capitalization collation test here?
e.g. This test validates that DAB creates queries with consistent capitalization of database object references (Title, TITLE, title)
or 
This test validates that DAB creates queries with all lower case letters (Title, TITLE -> title)
or
This test validates that DAB creates queries with casing consistent with the requirements of the backend database: database X has case sensitive requirements vs not.

File: src/Service.Tests/SqlTests/GraphQLCollationTests/PostgreSqlGraphQLCollationTests.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        [ClassInitialize]
+        public static async Task SetupAsync(TestContext context)
+        {
+            DatabaseEngine = TestCategory.POSTGRESQL;
+            await InitializeTestFixture();
+        }
+
+        #region Tests
+        /// <summary>
+        /// Postgre Capitalization Collation Tests
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/PostgreSqlGraphQLCollationTests.cs:25]:
```suggestion
        /// Postgres Capitalization Collation Tests
```

File: src/Service.Tests/SqlTests/GraphQLCollationTests/PostgreSqlGraphQLCollationTests.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        [DataTestMethod]
+        [DataRow("books", "title", @"SELECT json_agg(to_jsonb(table0)) FROM (SELECT title FROM books ORDER BY title asc) as table0")]
+        [DataRow("books", "title", @"SELECT json_agg(to_jsonb(table0)) FROM (SELECT TITLE FROM BOOKS ORDER BY TITLE asc) as table0")]
+        [DataRow("books", "title", @"SELECT json_agg(to_jsonb(table0)) FROM (SELECT Title FROM Books ORDER BY Title asc) as table0")]
+        public async Task PostgreCapitalizationResultQuery(string type, string item, string dbQuery)
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/PostgreSqlGraphQLCollationTests.cs:31]:
```suggestion
        public async Task PostgresCapitalizationResultQuery(string type, string item, string dbQuery)
```

File: src/Service.Tests/SqlTests/GraphQLCollationTests/GraphQLCollationTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        /// </summary>
+        public async Task CapitalizationResultQuery(string type, string item, string dbQuery)
```
Comment (seantleonard) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/GraphQLCollationTestBase.cs:18]:
```suggestion
        public async Task CapitalizationResultQuery(string objectType, string fieldName, string dbQuery)
```

more descriptive var names that align with GraphQL nomenclature per https://spec.graphql.org/October2021/#sec-Objects

**Review by seantleonard — APPROVED** [review]
Thanks for adding these tests. I suggest adding clarifying comments describing how the tests achieve their goals.

File: src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        /// </summary>
+        [DataTestMethod]
+        [DataRow("books", "title", @"SELECT title FROM books ORDER BY title ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        [DataRow("books", "title", @"SELECT title FROM BOOKS ORDER BY TITLE ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        [DataRow("books", "title", @"SELECT title FROM Books ORDER BY Title ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        public async Task MsSqlCapitalizationResultQuery(string type, string item, string dbQuery)
+        {
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs:34]: The test setup should change the Database collation. Then verify if DAB responds accurately irrespective of what collation the database is using. 

The tests proposed so far vary only in changing the capitalization of the expected database query. This does NOT represent what DAB does.

**Review by Aniruddh25 — CHANGES_REQUESTED** [review]
Need to understand what is under "testing"

File: src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs [hunk 8]
```diff
@@ -24,17 +28,29 @@ public static async Task SetupAsync(TestContext context)
...
-        public async Task MsSqlCapitalizationResultQuery(string type, string item, string dbQuery)
+        [DataRow("comics", "title", @"SELECT title FROM comics ORDER BY title ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        [DataRow("authors", "name", @"SELECT name FROM authors ORDER BY name ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        [DataRow("fungi", "habitat", @"SELECT habitat FROM fungi ORDER BY habitat ASC FOR JSON PATH, INCLUDE_NULL_VALUES")]
+        public async Task MsSqlCaseSensitiveResultQuery(string type, string item, string dbQuery)
         {
-            await CapitalizationResultQuery(type, item, dbQuery);
+            string defaultCollationQuery = MsSqlCollationQuery(type, item, DEFAULT_COLLATION);
+            string newCollationQuery = MsSqlCollationQuery(type, item, CASE_SENSITIVE_COLLATION);
+            await CaseSensitiveResultQuery(type, item, dbQuery, defaultCollationQuery, newCollationQuery);
```
Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs:41]: nit: rename type to objectType and item to fieldName in all the functions in all the classes...

Comment (Aniruddh25) [inline src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs:41]: nit: rename `CaseSensitiveResultQuery` to be a verb and better explain what the test is doing - e.g. TestQueryingWithCaseSensitiveCollation

**Review by Aniruddh25 — APPROVED** [review]
Looks good, just left some renames

File: src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs (hunk 1 above)
Comment (RubenCerna2079) [reply src/Service.Tests/SqlTests/GraphQLCollationTests/MsSqlGraphQLCollationTests.cs]: Changed tests, so this wouldn't apply anymore

**Pull request merged**